pop-api/src/error.rs

```diff
@@ -0,0 +1,255 @@
+				// TODO: make generic.
+				convert_to_fungibles_error(index, error)
+			} else {
+				error
+			}
+		},
+	};
+	ink::env::debug_println!("PopApiError: {:?}", error);
+	error
+}
+
+// If a unknown nested variant of the `DispatchError` is detected meaning any of the subsequent
+// bytes are non-zero (e.g. breaking change in the DispatchError), the error needs to be converted
+// into `PopApiError::Other`'s encoded value. This conversion is done by shifting the bytes one
+// position forward (the last byte is discarded as it is not being used) and replacing the first
+// byte with the `Other` encoded value (0u8). This ensures that the error is correctly categorized
+// as an `Other` variant.
+fn check_for_unknown_nesting(encoded_error: [u8; 4]) -> [u8; 4] {
+	if non_nested_pop_api_errors().contains(&encoded_error[0])
+		&& encoded_error[1..].iter().any(|x| *x != 0u8)
+	{
+		[0u8, encoded_error[0], encoded_error[1], encoded_error[2]]
+	} else if singular_nested_pop_api_errors().contains(&encoded_error[0])
+		&& encoded_error[2..].iter().any(|x| *x != 0u8)
+	{
+		[0u8, encoded_error[0], encoded_error[1], encoded_error[2]]
+	} else {
+		encoded_error
+	}
+}
+
+impl From<scale::Error> for PopApiError {
+	fn from(_: scale::Error) -> Self {
+		DecodingFailed
+	}
+}
+#[derive(Debug, Copy, Clone, PartialEq, Eq, scale::Encode, scale::Decode)]
+#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+pub enum TokenError {
+	/// Funds are unavailable.
+	FundsUnavailable,
+	/// Some part of the balance gives the only provider reference to the account and thus cannot
+	/// be (re)moved.
+	OnlyProvider,
+	/// Account cannot exist with the funds that would be given.
+	BelowMinimum,
+	/// Account cannot be created.
+	CannotCreate,
+	/// The asset in question is unknown.
+	UnknownAsset,
+	/// Funds exist but are frozen.
+	Frozen,
+	/// Operation is not supported by the asset.
+	Unsupported,
+	/// Account cannot be created for a held balance.
+	CannotCreateHold,
+	/// Withdrawal would cause unwanted loss of account.
+	NotExpendable,
+	/// Account cannot receive the assets.
+	Blocked,
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, scale::Encode, scale::Decode)]
+#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+pub enum ArithmeticError {
+	/// Underflow.
+	Underflow,
+	/// Overflow.
+	Overflow,
+	/// Division by zero.
+	DivisionByZero,
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, scale::Encode, scale::Decode)]
+#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+pub enum TransactionalError {
+	/// Too many transactional layers have been spawned.
+	LimitReached,
+	/// A transactional layer was expected, but does not exist.
+	NoLayer,
+}
+
+fn singular_nested_pop_api_errors() -> [u8; 3] {
+	const TOKEN_ERROR: u8 = 7;
+	const ARITHMETIC_ERROR: u8 = 8;
+	const TRANSACTION_ERROR: u8 = 9;
+	[TOKEN_ERROR, ARITHMETIC_ERROR, TRANSACTION_ERROR]
+}
+
+fn non_nested_pop_api_errors() -> [u8; 9] {
```

Comment:
Why define as a function rather than a constant?

Reply:
Very good point, changed to a constant